### ISSUE TITLE
Allowlist problems (Firefox 106, Shut Up 8.0)

### ISSUE THREAD

**Issue by philipmgrant** [RickyRomero/shut-up-webextension#35]
With the latest Firefox extension version (8.0):

a) When you add a site to the allowlist, comments don't appear until you refresh the page.

b) When you remove a site from the allowlist, comments don't disappear in the existing tab.

Steps to reproduce:

1. With no allowlist set up, open a Reddit thread or similar - the comments don't appear (as expected)

2. Click the toolbar button to allowlist the website.  The toolbar button changes appearance (the "-" sign appears) as expected, but comments still don't appear.

3. Hit F5 to refresh the page: now the comments appear.

4. Click the toolbar button again to remove the website from the allowlist.  The comments don't disappear.

   - Side note, often the toolbar button fails to change its appearance at this point (the "-" sign remains in place), but this doesn't seem related to the comments failing to disappear.  Comments are still shown whether or not the button changed appearance.

5. Refresh the page: the comments still appear despite the de-whitelisting.

6. Open the same page in a new tab: now the comments disappear.

I'm running Firefox 106.0b2 (Developer Edition) on Windows 10.

**Comment by RickyRomero** [comment]
Hi Philip! Sorry for the trouble, and I appreciate your detailed bug report.

This update made some deep changes to how the extension works, but it shouldn't have broken the toolbar button. I've gotten similar reports from a few other people.

I have a couple of suspicions for what might be the cause of this bug. Could you answer a couple of questions for me?:

1. I'm using a really fast development machine, and the update has quite a bit more asynchronous code in it, so a race condition might be one culprit. What are your PC's specs? And how many tabs do you usually have open in Firefox?
1. Since you're running the developer edition of Firefox, it's possible that Mozilla might've made some changes to default extension permissions. If you **right-click the toolbar icon** > choose **Manage Extension** > switch to the **Permissions** tab, what do you see there?

Let me know. Thanks!

**Comment by philipmgrant** [comment]
Thanks for your response!

1. Ryzen 5 2400G, 16GB RAM, with the Firefox profile on a RAMdisk.  Generally I have ~10 tabs open and FF is pretty responsive.

2. I see "Access your data for all websites" and "Access browser tabs": both with the green checkmark to show they have been granted.

**Comment by RickyRomero** [comment]
Thank you for the information! I'm working on some more changes right now. Would you be willing to help me test a development version later to see if it fixes the problem for you?

In the meantime, I will submit an update which restores the previous code for the time being.

**Comment by philipmgrant** [comment]
> Thank you for the information! I'm working on some more changes right now. Would you be willing to help me test a development version later to see if it fixes the problem for you?

For sure!

**Comment by Dio141** [comment]
this started happening again on firefox. same issue.

**Comment by RickyRomero** [comment]
@Dio141 I think that bug might be caused by something else, as I rolled the 8.0 changes back on Sep 21. I've gotten a few reports that this started happening after Firefox 106 was released. I'm looking into that now.

**Comment by RickyRomero** [comment]
@Dio141 I'm pretty sure I've found the source of the bug you reported. On Chrome and browsers based on it, I've set Shut Up to shut down during periods of low activity to save system resources.

https://github.com/RickyRomero/shut-up-webextension/blob/d4c818e17f9df2216d98014566e2455813391737/manifest.json#L48

Firefox simply ignored this manifest flag [until version 106 came out.](https://developer.mozilla.org/en-US/docs/Mozilla/Firefox/Releases/106#changes_for_add-on_developers)

This really shouldn't have caused a problem since Chrome handles this flag gracefully, but Firefox does not appear to implement it in the same way. This causes the extension to break for an unknown reason after its process is stopped.

I'm submitting an update for the Firefox version tonight to force the extension back into the old persistent mode, which should fix the bug you encountered. After that, I'll see if it's possible for me to use the ephemeral mode correctly in Firefox.

**Comment by RickyRomero** [comment]
@Dio141 I've released version 8.0.2 for Firefox (based on 7.1.2) which should address the problem you encountered. Let me know how it goes.

**Comment by Dio141** [comment]
seems to be working perfectly now!

**Comment by philipmgrant** [comment]
All looking good for me too!  Thanks for your work on this!

**Comment by RickyRomero** [comment]
I'm pretty sure this bug is fixed in version 8.1 (which is based on the code from the 8.0 release this bug originated from). The background script isn't persistent in either browser anymore and it seems to work fine. I'm pushing the update this month. If you see the problem appear again after the update, just bump this ticket and I'll reopen it.

Thanks again for the bug report!

**Comment by philipmgrant** [comment]
Unfortunately I've been seeing a related but not identical issue since my Firefox extension updated to 8.1 today.

When navigating to a domain which had been added to the allowlist, comments are blocked and the toolbar button remains in its normal style without the "-" sign.

Clicking the toolbar button results in the button showing the "-" sign, and the comments being shown (not like the original issue, where a further refresh was needed before comments would appear.)

However, this preference doesn't stick. Navigating to a different page in the same allowlisted domain (even in the same tab where I clicked the button and saw the comments appear) results in comments disappearing again.

**Comment by RickyRomero** [comment]
@philipmgrant Thanks for letting me know. Can you take a look at the 8.1.1 update I just released? I worked around a serious bug in Firefox's `insertCSS`/`removeCSS` APIs.

**Comment by philipmgrant** [comment]
@RickyRomero Thanks very much!

I realised that the problems I reported with 8.1 may have had a very simple cause: it looks like updating to 8.1 had caused the "Remember where I've enabled comments" setting to revert to False. Setting that to True fixed the behaviour that I mentioned.

However, on 8.1 I was also seeing an additional issue where sometimes the extension toolbar icon would be greyed out and comments would not be hidden. Since updating to 8.1.1 and restarting, I haven't seen a recurrence of that. So it's looking good so far!
